src/System.Text.Json/src/System/Text/Json/JsonDocument.CustomDb.cs

```diff
@@ -0,0 +1,206 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+namespace System.Text.Json
+{
+    partial class JsonDocument
+    {
+        private struct CustomDb : IDisposable
+        {
+            private const int SizeOrLengthOffset = 4;
+            private const int NumberOfRowsOffset = 8;
+
+            internal int Length;
+            private byte[] _rentedBuffer;
+            private ArrayPool<byte> _pool;
+
+            internal CustomDb(ArrayPool<byte> pool, int initialSize)
+            {
+                _pool = pool;
+                _rentedBuffer = _pool.Rent(initialSize);
+                Length = 0;
+            }
+
+            public void Dispose()
+            {
+                if (_rentedBuffer == null)
+                {
+                    return;
+                }
+
+                _pool.Return(_rentedBuffer);
+                _rentedBuffer = null;
```

Comment:
We obviously need to be really careful with this... we could easily end up returning the same buffer to the pool multiple times if Dispose were called on copies of the same struct.